codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/DelegatingSyncClientClass.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.codegen.poet.client;
+
+import static software.amazon.awssdk.codegen.internal.Constant.EVENT_PUBLISHER_PARAM_NAME;
+import static software.amazon.awssdk.codegen.internal.Constant.EVENT_RESPONSE_HANDLER_PARAM_NAME;
+import static software.amazon.awssdk.codegen.internal.Constant.SYNC_STREAMING_INPUT_PARAM;
+import static software.amazon.awssdk.codegen.internal.Constant.SYNC_STREAMING_OUTPUT_PARAM;
+
+import com.squareup.javapoet.ClassName;
+import com.squareup.javapoet.FieldSpec;
+import com.squareup.javapoet.MethodSpec;
+import com.squareup.javapoet.TypeSpec;
+import java.util.List;
+import javax.lang.model.element.Modifier;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.codegen.docs.SimpleMethodOverload;
+import software.amazon.awssdk.codegen.model.config.customization.UtilitiesMethod;
+import software.amazon.awssdk.codegen.model.intermediate.IntermediateModel;
+import software.amazon.awssdk.codegen.model.intermediate.OperationModel;
+import software.amazon.awssdk.codegen.poet.PoetExtension;
+import software.amazon.awssdk.codegen.poet.PoetUtils;
+import software.amazon.awssdk.codegen.utils.PaginatorUtils;
+
+public class DelegatingSyncClientClass extends SyncClientInterface {
+
+    private final IntermediateModel model;
+    private final ClassName className;
+    private final PoetExtension poetExtensions;
+
+    public DelegatingSyncClientClass(IntermediateModel model) {
+        super(model);
+        this.model = model;
+        this.className = ClassName.get(model.getMetadata().getFullClientPackageName(),
+                                       "Delegating" + model.getMetadata().getSyncInterface());
+        this.poetExtensions = new PoetExtension(model);
+    }
+
+    @Override
+    public TypeSpec poetSpec() {
+        ClassName interfaceClass = poetExtensions.getClientClass(model.getMetadata().getSyncInterface());
+        TypeSpec.Builder result = PoetUtils.createClassBuilder(className);
+
+        result.addSuperinterface(interfaceClass)
+              .addAnnotation(SdkProtectedApi.class)
+              .addModifiers(Modifier.ABSTRACT, Modifier.PUBLIC)
```

Comment:
I think all methods are implemented except serviceName. One benefit of using abstract is that people can't initialize it directly, which is not the intended use-case.

Reply:
serviceName() in SdkClient is unimplemented

Reply:
Any reason not to implement `serviceName()`? I feel like it shouldn't differ in subclasses?

Also, should we create a subclass of the delegating class in our codegen generated classes test module, so that if we fail to override a method, it'll fail compilation in our project instead of in our customer's?

Reply:
implemented serviceName in DelegatingSyncClientClass